compiler/src/vm/tracer/mod.rs

```diff
@@ -0,0 +1,343 @@
+pub mod stack_trace;
+
+use super::{heap::Pointer, ChannelId, FiberId, Heap};
+use crate::{compiler::hir::Id, module::Module};
+use itertools::Itertools;
+use std::{collections::HashMap, fmt, time::Instant};
+
+pub trait Tracer {
+    fn fiber_created(&mut self, fiber: FiberId);
+    fn fiber_done(&mut self, fiber: FiberId);
+    fn fiber_panicked(&mut self, fiber: FiberId, panicked_child: Option<FiberId>);
+    fn fiber_canceled(&mut self, fiber: FiberId);
+    fn fiber_execution_started(&mut self, fiber: FiberId);
+    fn fiber_execution_ended(&mut self, fiber: FiberId);
+    fn channel_created(&mut self, channel: ChannelId);
+    fn sent_to_channel(&mut self, value: Pointer, from: FiberId, to: ChannelId);
+    fn received_from_channel(&mut self, value: Pointer, from: ChannelId, to: FiberId);
+
+    fn in_fiber_tracer<'a>(&'a mut self, fiber: FiberId) -> Box<dyn InFiberTracer<'a> + 'a>
+    where
+        Self: 'a;
+}
+
+pub trait InFiberTracer<'a> {
+    fn module_started(&mut self, module: Module);
+    fn module_ended(&mut self, heap: &Heap, export_map: Pointer);
+    fn value_evaluated(&mut self, heap: &Heap, id: Id, value: Pointer);
+    fn found_fuzzable_closure(&mut self, heap: &Heap, id: Id, closure: Pointer);
+    fn call_started(&mut self, heap: &Heap, id: Id, closure: Pointer, args: Vec<Pointer>);
+    fn call_ended(&mut self, heap: &Heap, return_value: Pointer);
+    fn needs_started(&mut self, heap: &Heap, id: Id, condition: Pointer, reason: Pointer);
+    fn needs_ended(&mut self);
+}
+
+// A dummy version of the tracer that is used when running known instructions
+// without wanting to trace them.
+
+pub struct DummyTracer;
+pub struct DummyInFiberTracer;
+impl Tracer for DummyTracer {
+    fn fiber_created(&mut self, _fiber: FiberId) {}
```

Comment:
Since we have the `Event` enum anyway, maybe we should simplify the `Tracer` to just have a single method accepting an `Event`? That should save quite a few lines in the `Tracer` implementations

Reply:
Hmm, maybe. I definitely agree that there is some boilerplate code when implementing a tracer. But when designing the tracer, I first thought about what an optimal usage API would look like and then implemented that.

I feel like while taking an `Event` makes the tracer implementations themselves cleaner, it pushes some complexity into the call sites. For example, _some_ methods also take a `Heap` so they can clone objects into the tracer. If we create a general method that accepts an `Event`, that would also need to accept a heap.

I actually started doing the work for it last night, but there are some ugly changes to the caller locations:

```rust
tracer.fiber_execution_started(fiber_id);
// becomes:
tracer.add(
  Event::FiberExecutionStarted { fiber: fiber_id },
  &Heap::new(),
);
```

Actually, now that I think about it, what do you think about splitting the tracer further up into a `Tracer` and `InFiberTracer` with an optimal API and a `TracerImplementation` which only has a single method accepting an `Event` and an `Option<&Heap>`? This way, we would have both a clean API as well as simple implementations.